### PR TITLE
CI: In the `asan` CI job, enable both Julia assertions and LLVM assertions

### PR DESCRIPTION
**Pull request by DilumAluthge**
See https://github.com/JuliaLang/julia/issues/44361#issuecomment-1053134163 and https://github.com/JuliaLang/julia/issues/44361#issuecomment-1053144552 for context.

### CODE REVIEW

File: contrib/asan/build.sh [hunk 1]
```diff
@@ -34,21 +34,25 @@ JULIA_HOME="$HERE/../../"
 echo
 echo "Installing toolchain..."
 
+# Enable Julia assertions: FORCE_ASSERTIONS=1
+# Enable LLVM assertions:  LLVM_ASSERTIONS=1
+MAKE_ASSERT_FLAGS="FORCE_ASSERTIONS=1 LLVM_ASSERTIONS=1"
```
Comment (tkf) [inline contrib/asan/build.sh:39]: Maybe move this to `Make.user.asan`? (and `Make.user.tools`?)

Comment (DilumAluthge) [reply contrib/asan/build.sh:39]: Done! I've added it to both `Make.user.asan` and `Make.user.tools`.

Should I now remove it from `build.sh`?

Comment (DilumAluthge) [reply contrib/asan/build.sh]: I've removed it from `build.sh`.

**Comment by DilumAluthge** [comment]
@tkf Can you reproduce these error messages locally?

**Comment by vtjnash** [comment]
Yggdrasil does not support clang+LLVM_ASSERTIONS: https://github.com/JuliaLang/julia/blob/e84b06c67393ee10799535781144de3c2d8860b7/deps/llvm.mk#L297-L303

If you need to build with a copy of `clang`, then you can set do a partial build of just clang:
```
make configure O=just-clang
make -C just-clang/deps install-clang
export CC=just-clang/usr/tools/clang
export CXX=just-clang/usr/tools/clang++
```

**Comment by DilumAluthge** [comment]
Oh that's interesting. Does that mean that you can't do a Julia assertions + LLVM assertions build on macOS?

File: contrib/asan/Make.user.tools [hunk 2]
```diff
@@ -1,2 +1,6 @@
 USE_BINARYBUILDER_LLVM=1
 BUILD_LLVM_CLANG=1
+
+# Enable Julia assertions and LLVM assertions
+FORCE_ASSERTIONS=1
+LLVM_ASSERTIONS=1
```
Comment (vtjnash) [inline contrib/asan/Make.user.tools:6]:
```suggestion
```
You don't need to install `clang` with julia assertions enabled.

Comment (DilumAluthge) [reply contrib/asan/Make.user.tools]: Oh, is the issue specifically when trying to build a new clang from source with LLVM assertions?

Comment (vtjnash) [reply contrib/asan/Make.user.tools]: This file does not build llvm from source

Comment (DilumAluthge) [reply contrib/asan/Make.user.tools]: What is this file used for?

**Comment by DilumAluthge** [comment]
It looks like it's working!

> julia-debug: /workspace/srcdir/llvm-project/llvm/include/llvm/IR/Type.h:461: static llvm::Type* llvm::Type::getFloatingPointTy(llvm::LLVMContext&, const llvm::fltSemantics&): Assertion `&S == &APFloat::PPCDoubleDouble() && "Unknown FP format"' failed.

**Comment by DilumAluthge** [comment]
Now someone just needs to fix the actual bug 😂 😂 😂

**Pull request merged**